django/db/models/fields/related.py

```diff
@@ -26,74 +28,53 @@
+      * A model class, which will be returned unchanged.
     """
     # Check for recursive relations
     if relation == RECURSIVE_RELATIONSHIP_CONSTANT:
-        app_label = cls._meta.app_label
-        model_name = cls.__name__
+        relation = scope_model
 
-    else:
-        # Look for an "app.Model" relation
+    # Look for an "app.Model" relation
+    if isinstance(relation, six.string_types):
+        if "." not in relation:
+            relation = "%s.%s" % (scope_model._meta.app_label, relation)
 
-        if isinstance(relation, six.string_types):
-            try:
-                app_label, model_name = relation.split(".")
-            except ValueError:
-                # If we can't split, assume a model in current app
-                app_label = cls._meta.app_label
-                model_name = relation
-        else:
-            # it's actually a model class
-            app_label = relation._meta.app_label
-            model_name = relation._meta.object_name
-
-    # Try to look up the related model, and if it's already loaded resolve the
-    # string right away. If get_registered_model raises a LookupError, it means
-    # that the related model isn't loaded yet, so we need to pend the relation
-    # until the class is prepared.
-    try:
-        model = cls._meta.apps.get_registered_model(app_label, model_name)
-    except LookupError:
-        key = (app_label, model_name)
-        value = (cls, field, operation)
-        cls._meta.apps._pending_lookups.setdefault(key, []).append(value)
-    else:
-        operation(field, model, cls)
-
-
-def do_pending_lookups(sender, **kwargs):
+    return relation
+
+
+def lazy_related_operation(function, model, *related_models, **kwargs):
     """
-    Handle any pending relations to the sending model. Sent from class_prepared.
+    A wrapper for Apps.lazy_model_operation() where the first model has some
+    special significance:
+      * Uses the apps object from the first model argument's _meta
+      * Accepts as its positional arguments the string "self" or model names
+        without an app label, which will be resolved into a fully-qualified
+        model name relative to the first model using resolve_relation().
     """
-    key = (sender._meta.app_label, sender.__name__)
-    for cls, field, operation in sender._meta.apps._pending_lookups.pop(key, []):
-        operation(field, sender, cls)
+    apps = model._meta.apps
+    related_models = [resolve_relation(model, rel) for rel in related_models]
+    return apps.lazy_model_operation(partial(function, **kwargs), model, *related_models)
 
-signals.class_prepared.connect(do_pending_lookups)
+
+def add_lazy_relation(cls, field, relation, operation):
+    warnings.warn(
+        "add_lazy_relation() has been replaced by lazy_related_operation() "
+        "and related methods on the Apps class, and will be no longer be "
+        "available in Django 1.9.", RemovedInDjango19Warning, stacklevel=2)
```

Comment:
Assuming this patch is only going into Django 1.9, I think this should be a `RemovedInDjango21Warning`, unless we are doing an accelerated deprecation because it's an undocumented internal API?